Limit hostnames to 31 characters

oxan opened the pull request:
# What does this implement/fix? 

Both `tcpip_adapter_set_hostname` on the ESP32 and `wifi_station_set_hostname` on the ESP8266 don't accept hostnames longer than 32 bytes (so 31 characters). Limit the length of the hostname to 31 characters. 

Alternatively we could warn and truncate the hostname, but I don't think many people are using hostnames that are this long.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

oxan committed:
Limit hostnames to 31 characters

OttoWinter reviewed (approved):
:+1:

We should also add some validation to reduce the length by 9 (?) if `name_add_mac_suffix` is true

oxan committed:
Reduce limit when add_mac_suffix is used

oxan commented:
Okay, rewrote the PR to use a reduced limit when `name_add_mac_suffix` is enabled. Moved the validation of our hostname to `core/config.py`, `cv.hostname` can now be used to check whether an external hostname is valid.

oxan committed:
Add missing return
